contrib/labstack/echo.v4/appsec.go

```diff
@@ -6,33 +6,52 @@
 package echo
 
 import (
+	"net/http"
+
 	"gopkg.in/DataDog/dd-trace-go.v1/ddtrace/tracer"
-	"gopkg.in/DataDog/dd-trace-go.v1/internal/appsec/dyngo/instrumentation"
 	"gopkg.in/DataDog/dd-trace-go.v1/internal/appsec/dyngo/instrumentation/httpsec"
 
 	"github.com/labstack/echo/v4"
 )
 
-func useAppSec(c echo.Context, span tracer.Span) func() {
-	instrumentation.SetAppSecEnabledTags(span)
-
+func withAppSec(next echo.HandlerFunc, c echo.Context, span tracer.Span) echo.HandlerFunc {
 	params := make(map[string]string)
 	for _, n := range c.ParamNames() {
 		params[n] = c.Param(n)
 	}
```

Comment:
For the sake of correctness, by making this function an echo middleware you no longer want `c echo.Context` as argument anymore, because you are not even supposed to have it at "handler initialization time".

The fact we use this function from within the request handler makes it available, but still I think you should follow the expected "middleware api" and especially avoid having the request context twice like it is the case here (one as middlware arg, the other in the returned handler args).

This also means the params creation needs to move into the return request handler.

Reply:
Agreed for the double context.
Normalizing to the echo middleware API also means retrieving the span from the context in the hot path rather than at handler initialization (on top of  path params creation like you mentioned)